### ISSUE TITLE
ChildIOException when re-running a pipeline that create links to files in a directory output

### ISSUE THREAD

**Issue by blaiseli** [snakemake/snakemake#413]
**Snakemake version**
5.18.1

**Describe the bug**
I have a workflow with two rules:

Rule `make_test_dir_and_test_file` creates a directory, a subdirectory, and a file within that subdirectory:

```python
rule make_test_dir_and_test_file:
    output:
        directory("test_dir/subdir"),
    shell:
        """
        mkdir -p {output[0]}/
        echo "test" > {output[0]}/test_file
        """
```

The other rule, `make_other_file`, takes as input the directory created by `make_test_dir_and_test_file` and creates in the "main" directory (the parent of its input) a link to the contained file:

```python
rule make_other_file:
    input:
        "test_dir/subdir"
    output:
        "test_dir/other_file"
    shell:
        """
        ln -s subdir/test_file {output[0]}
        """
```

This link is my final target:

```python
rule all:
    input:
        "test_dir/other_file"
```

I can run the workflow once successfully, but when I try to re-run it, I get a `ChildIOException` that tells me that the linked file is a child of it's containing grand-parent directory:
```
ChildIOException:
File/directory is a child to another output:
/home/bli/src/10825_Cecere/Genomes/test_dir/subdir
/home/bli/src/10825_Cecere/Genomes/test_dir/subdir/test_file
```

I do not expect this to happen, because I'm not creating this file in a separate rule: this file was created in the same rule as its containing directory. What I create in the second rule is a link, which resides in another directory (the main `test_dir`).

With another workflow where I remove one level in the file hierarchy (the `subdir` part on the paths), I get a `ChildIOException` right away, but that seems expected, because the link created in the second rule (`test_dir/other_file`) is now indeed a child of the directory created by the first rule (`test_dir`).

**Logs**
```
# Clean output
$ rm -rf test_dir/
$ snakemake --cores 1 --snakefile test4.snakefile 2> log
# Did it work?
$ echo $?
0
$ snakemake --cores 1 --snakefile test4.snakefile 2> log
# Did it work again?
$ echo $?
1
$ cat log 
Building DAG of jobs...
# Re-works after cleaning the output:
$ rm -rf test_dir/
$ snakemake --cores 1 --snakefile test4.snakefile 2> log
$ echo $?
0
```

**Minimal example**

Here is `test4.snakefile`:

```python
rule all:
    input:
        "test_dir/other_file"

rule make_test_dir_and_test_file:
    output:
        directory("test_dir/subdir"),
    shell:
        """
        mkdir -p {output[0]}/
        echo "test" > {output[0]}/test_file
        """

rule make_other_file:
    input:
        "test_dir/subdir"
    output:
        "test_dir/other_file"
    shell:
        """
        ln -s subdir/test_file {output[0]}
        """
```

**Additional context**
I'm using python 3.8.2 on a Linux workstation under MX Linux, a Debian-based distribution:
```
$ python --version
Python 3.8.2
$ uname -a
Linux pisa 4.19.0-5-amd64 #1 SMP Debian 4.19.37-2~mx17+1 (2019-05-15) x86_64 GNU/Linux
```

This issue follows this one:
https://github.com/snakemake/snakemake/issues/390

**Comment by blaiseli** [comment]
The issue might be due to the use of `os.path.abspath`, which looks at the linked file instead of the link. I suppose there is a good reason to do that, but I don't understand it.

**Comment by jmeppley** [comment]
I'm seeing this, too.

It does not happen in 5.17 or 5.18.1.

It does occur in 5.19.0 and 5.19.3 and the latest code (5.19.3+1.g7026550f, last commit: 7026550fabcdfe952029b75b34cd583371eb7f4c)

Here is my minimal Snakefile:

```
rule link_file:
    input: 'real_file.txt'
    output: 'symbolic_file.txt'
    shell: 'ln -s real_file.txt {output}'

rule first_rule:
    output: 'real_file.txt'
    shell: 'echo DATA > {output}'
```

I'm using python 3.8.3 on Fedora 30.

```
(/mnt/histidine/scratch2/jmeppley/tmp/snakemake_bug/env.dev) [jmeppley@tyrosine snakemake_bug]$ snakemake -j
Building DAG of jobs...
ChildIOException:
File/directory is a child to another output:
('/mnt/histidine/scratch2/jmeppley/tmp/snakemake_bug/real_file.txt', first_rule)
('/mnt/histidine/scratch2/jmeppley/tmp/snakemake_bug/real_file.txt', link_file)
(/mnt/histidine/scratch2/jmeppley/tmp/snakemake_bug/env.dev) [jmeppley@tyrosine snakemake_bug]$ snakemake --version
5.19.3+1.g7026550f
(/mnt/histidine/scratch2/jmeppley/tmp/snakemake_bug/env.dev) [jmeppley@tyrosine snakemake_bug]$ conda deactivate
(base) [jmeppley@tyrosine snakemake_bug]$ conda activate ./env18
(/mnt/histidine/scratch2/jmeppley/tmp/snakemake_bug/env18) [jmeppley@tyrosine snakemake_bug]$ snakemake --version
5.18.1
(/mnt/histidine/scratch2/jmeppley/tmp/snakemake_bug/env18) [jmeppley@tyrosine snakemake_bug]$ snakemake -j
Building DAG of jobs...
Nothing to be done.
Complete log: /mnt/histidine/scratch2/jmeppley/tmp/snakemake_bug/.snakemake/log/2020-06-17T135756.510504.snakemake.log
```

The problem arises, I think, because dag.py is using os.path.realpath to compare outputs.

**Comment by jmeppley** [comment]
The bug was introduced with this change:

```
 [jmeppley@tyrosine snakemake]$ git diff 0c289a650803f08f35e68392fb066e8977950b2b f5adeaa434b0f207972ceb24889313d8bb7df444
diff --git a/snakemake/dag.py b/snakemake/dag.py
index b1c9317e..226a5d02 100755
--- a/snakemake/dag.py
+++ b/snakemake/dag.py
@@ -189,20 +189,20 @@ class DAG:
         """Check that no output file is contained in a directory output of the same or another rule."""
         outputs = sorted(
             {
-                (path(f), job)
+                path(f)
                 for job in self.jobs
                 for f in job.output
                 for path in (os.path.abspath, os.path.realpath)
             }
         )
         for i in range(len(outputs) - 1):
-            (a, job_a), (b, job_b) = outputs[i : i + 2]
+            a, b = outputs[i : i + 2]
             try:
                 common = os.path.commonpath([a, b])
             except ValueError:
                 # commonpath raises error if windows drives are different.
                 continue
-            if common == os.path.commonpath([a]) and job_a != job_b:
+            if common == os.path.commonpath([a]):
                 raise ChildIOException(parent=outputs[i], child=outputs[i + 1])
 ```

Previously, since the link resolves to the SAME path as the original file via realpath, they were combined into entr in the set(). Now, by changing from path(f) to (path(f), job), they are two different entries.

**Comment by jmeppley** [comment]
The following solves this bug, but I'm not sure if it introduces others:

```
diff --git a/snakemake/dag.py b/snakemake/dag.py
index 20b857b3..acab7e2a 100755
--- a/snakemake/dag.py
+++ b/snakemake/dag.py
@@ -207,7 +207,7 @@ class DAG:
             except ValueError:
                 # commonpath raises error if windows drives are different.
                 continue
-            if common == os.path.commonpath([a]) and job_a != job_b:
+            if common == os.path.commonpath([a]) and job_a != job_b and a != b:
                 raise ChildIOException(parent=outputs[i], child=outputs[i + 1])
```

**Comment by jmeppley** [comment]
The other option is to drop os.path.realpath. I don't see the reason for its inclusion. It originally was introduced in commit 07f0451aa9c19fac4de36ad6c5ce5d998d3b330d when inputs were compared, too. Admittedly, I don't know the full scope of things here, but I think we can just use abspath for this:

```
diff --git a/snakemake/dag.py b/snakemake/dag.py
index 20b857b3..d7c94579 100755
--- a/snakemake/dag.py
+++ b/snakemake/dag.py
@@ -194,10 +194,9 @@ class DAG:
         """Check that no output file is contained in a directory output of the same or another rule."""
         outputs = sorted(
             {
-                (path(f), job)
+                (os.path.abspath(f), job)
                 for job in self.jobs
                 for f in job.output
-                for path in (os.path.abspath, os.path.realpath)
             }
         )
         for i in range(len(outputs) - 1):
```

**Comment by jmeppley** [comment]
I prefer the second solution to the first because it simplifies the logic rather than adding to it. The only catch would be if it now lets through conflicting outputs that it shouldn't.

**Comment by blaiseli** [comment]
I was also trying to fix this (https://github.com/snakemake/snakemake/pull/452)

My approach looks much more complicated (and I haven't actually checked it solves the issue): I separated parents and childs in the outputs, with realpath kept when considering an output as parent, and removed when considered as a child. The idea is that when creating a child, what matters is the actual file created (real or link), and not the target (if the file is a link). This separation lead to a complexification of the logic.

**Comment by VGalata** [comment]
I have the same issue with the version `5.20.1`.

**Comment by tedtoal** [comment]
This issue doesn't appear to actually have anything to do with directories, only with links to files.  Here's my snake file:

```
rule all:
	input: "test1.txt"

rule A:
	output: "test1.txt"
	input: "test.txt"
	shell: "ln -s test.txt {output}"

rule B:
	output: "test.txt"
	shell: "echo 'rule B' >{output}"
```

The first time it runs just fine and creates text file test.txt and a link to it named test1.txt.  If it is run again, I get the error:

```
ChildIOException:
File/directory is a child to another output:
('/Users/tedtoal/Documents/UCDavis/Carvajal/PipelineTesting/snakemake/test.txt', B)
('/Users/tedtoal/Documents/UCDavis/Carvajal/PipelineTesting/snakemake/test.txt', A)
```

I have to go back to version 5.17.0 for this error to not happen.

**Comment by jmeppley** [comment]
Thanks, @tedtoal, That agrees with my experience, except I had 5.18.1 working. 

There are two PRs under consideration for this. I'm not sure when they'll get in the official code. If you need the latest snakemake to work with symlinks, you should be able to use PR #447. It works for me and has passed all the automated checks.

**Comment by tedtoal** [comment]
I'll just stick with an older version for now.

**Comment by VGalata** [comment]
Could we get an update on this issue? It's a major problem for anyone using symlinks.

**Comment by johanneskoester** [comment]
Should be fixed with PR #825. Thanks a lot @jmeppley!
